File: autoimageslider/src/main/java/com/smarteist/autoimageslider/IndicatorView/PageIndicatorView.java
package com.smarteist.autoimageslider.IndicatorView;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.database.DataSetObserver;
import android.graphics.Canvas;
import android.os.Build;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.text.TextUtilsCompat;
import androidx.viewpager.widget.PagerAdapter;
import androidx.core.view.ViewCompat;
import androidx.viewpager.widget.ViewPager;
import android.util.AttributeSet;
import android.util.Pair;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import com.smarteist.autoimageslider.IndicatorView.animation.type.AnimationType;
import com.smarteist.autoimageslider.IndicatorView.animation.type.ScaleAnimation;
import com.smarteist.autoimageslider.IndicatorView.draw.controller.DrawController;
import com.smarteist.autoimageslider.IndicatorView.draw.data.Indicator;
import com.smarteist.autoimageslider.IndicatorView.draw.data.Orientation;
import com.smarteist.autoimageslider.IndicatorView.draw.data.PositionSavedState;
import com.smarteist.autoimageslider.IndicatorView.draw.data.RtlMode;
import com.smarteist.autoimageslider.IndicatorView.utils.CoordinatesUtils;
import com.smarteist.autoimageslider.IndicatorView.utils.DensityUtils;
import com.smarteist.autoimageslider.IndicatorView.utils.IdUtils;

public class PageIndicatorView extends View implements ViewPager.OnPageChangeListener, IndicatorManager.Listener, ViewPager.OnAdapterChangeListener {

    private IndicatorManager manager;
    private DataSetObserver setObserver;
    private ViewPager viewPager;
    private boolean isInteractionEnabled;

    public PageIndicatorView(Context context) {
        super(context);
        init(null);
    }

    public PageIndicatorView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public PageIndicatorView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public PageIndicatorView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(attrs);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        findViewPager(getParent());
    }

    @Override
    protected void onDetachedFromWindow() {
        unRegisterSetObserver();
        super.onDetachedFromWindow();
    }

    @Override
    public Parcelable onSaveInstanceState() {
        Indicator indicator = manager.indicator();
        PositionSavedState positionSavedState = new PositionSavedState(super.onSaveInstanceState());
        positionSavedState.setSelectedPosition(indicator.getSelectedPosition());
        positionSavedState.setSelectingPosition(indicator.getSelectingPosition());
        positionSavedState.setLastSelectedPosition(indicator.getLastSelectedPosition());

        return positionSavedState;
    }

    @Override
    public void onRestoreInstanceState(Parcelable state) {
        if (state instanceof PositionSavedState) {
            Indicator indicator = manager.indicator();
            PositionSavedState positionSavedState = (PositionSavedState) state;
            indicator.setSelectedPosition(positionSavedState.getSelectedPosition());
            indicator.setSelectingPosition(positionSavedState.getSelectingPosition());
            indicator.setLastSelectedPosition(positionSavedState.getLastSelectedPosition());
            super.onRestoreInstanceState(positionSavedState.getSuperState());

        } else {
            super.onRestoreInstanceState(state);
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        Pair<Integer, Integer> pair = manager.drawer().measureViewSize(widthMeasureSpec, heightMeasureSpec);
        setMeasuredDimension(pair.first, pair.second);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        manager.drawer().draw(canvas);
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        manager.drawer().touch(event);
        return true;
    }

    @Override
    public void onIndicatorUpdated() {
        invalidate();
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        onPageScroll(position, positionOffset);
    }

    @Override
    public void onPageSelected(int position) {
        onPageSelect(position);
    }

    @Override
    public void onPageScrollStateChanged(int state) {
		if (state == ViewPager.SCROLL_STATE_IDLE) {
			manager.indicator().setInteractiveAnimation(isInteractionEnabled);
		}
	}

    @Override
    public void onAdapterChanged(@NonNull ViewPager viewPager, @Nullable PagerAdapter oldAdapter, @Nullable PagerAdapter newAdapter) {
        updateState();
    }


    public void setCount(int count) {
        if (count >= 0 && manager.indicator().getCount() != count) {
            manager.indicator().setCount(count);
            updateVisibility();
            requestLayout();
        }
    }


    public int getCount() {
        return manager.indicator().getCount();
    }


    public void setDynamicCount(boolean dynamicCount) {
        manager.indicator().setDynamicCount(dynamicCount);

        if (dynamicCount) {
            registerSetObserver();
        } else {
            unRegisterSetObserver();
        }
    }


    public void setRadius(int radiusDp) {
        if (radiusDp < 0) {
            radiusDp = 0;
        }

        int radiusPx = DensityUtils.dpToPx(radiusDp);
        manager.indicator().setRadius(radiusPx);
        invalidate();
    }


    public void setRadius(float radiusPx) {
        if (radiusPx < 0) {
            radiusPx = 0;
        }

        manager.indicator().setRadius((int) radiusPx);
        invalidate();
    }


    public int getRadius() {
        return manager.indicator().getRadius();
    }


    public void setPadding(int paddingDp) {
        if (paddingDp < 0) {
            paddingDp = 0;
        }

        int paddingPx = DensityUtils.dpToPx(paddingDp);
        manager.indicator().setPadding(paddingPx);
        invalidate();
    }


    public void setPadding(float paddingPx) {
        if (paddingPx < 0) {
            paddingPx = 0;
        }

        manager.indicator().setPadding((int) paddingPx);
        invalidate();
    }


    public int getPadding() {
        return manager.indicator().getPadding();
    }


    public void setScaleFactor(float factor) {
        if (factor > ScaleAnimation.MAX_SCALE_FACTOR) {
            factor = ScaleAnimation.MAX_SCALE_FACTOR;

        } else if (factor < ScaleAnimation.MIN_SCALE_FACTOR) {
            factor = ScaleAnimation.MIN_SCALE_FACTOR;
        }

        manager.indicator().setScaleFactor(factor);
    }


    public float getScaleFactor() {
        return manager.indicator().getScaleFactor();
    }


    public void setStrokeWidth(float strokePx) {
        int radiusPx = manager.indicator().getRadius();

        if (strokePx < 0) {
            strokePx = 0;

        } else if (strokePx > radiusPx) {
            strokePx = radiusPx;
        }

        manager.indicator().setStroke((int) strokePx);
        invalidate();
    }



    public void setStrokeWidth(int strokeDp) {
        int strokePx = DensityUtils.dpToPx(strokeDp);
        int radiusPx = manager.indicator().getRadius();

        if (strokePx < 0) {
            strokePx = 0;

        } else if (strokePx > radiusPx) {
            strokePx = radiusPx;
        }

        manager.indicator().setStroke(strokePx);
        invalidate();
    }


    public int getStrokeWidth() {
        return manager.indicator().getStroke();
    }


    public void setSelectedColor(int color) {
        manager.indicator().setSelectedColor(color);
        invalidate();
    }


    public int getSelectedColor() {
        return manager.indicator().getSelectedColor();
    }


    public void setUnselectedColor(int color) {
        manager.indicator().setUnselectedColor(color);
        invalidate();
    }


    public int getUnselectedColor() {
        return manager.indicator().getUnselectedColor();
    }


    public void setAutoVisibility(boolean autoVisibility) {
        if (!autoVisibility) {
            setVisibility(VISIBLE);
        }

        manager.indicator().setAutoVisibility(autoVisibility);
        updateVisibility();
    }


    public void setOrientation(@Nullable Orientation orientation) {
        if (orientation != null) {
            manager.indicator().setOrientation(orientation);
            requestLayout();
        }
    }


    public void setAnimationDuration(long duration) {
        manager.indicator().setAnimationDuration(duration);
    }


    public long getAnimationDuration() {
        return manager.indicator().getAnimationDuration();
    }


    public void setAnimationType(@Nullable AnimationType type) {
        manager.onValueUpdated(null);

        if (type != null) {
            manager.indicator().setAnimationType(type);
        } else {
            manager.indicator().setAnimationType(AnimationType.NONE);
        }
        invalidate();
    }


    public void setInteractiveAnimation(boolean isInteractive) {
        manager.indicator().setInteractiveAnimation(isInteractive);
        this.isInteractionEnabled = isInteractive;
    }


    public void setViewPager(@Nullable ViewPager pager) {
        releaseViewPager();
        if (pager == null) {
            return;
        }

        viewPager = pager;
        viewPager.addOnPageChangeListener(this);
        viewPager.addOnAdapterChangeListener(this);
        manager.indicator().setViewPagerId(viewPager.getId());

        setDynamicCount(manager.indicator().isDynamicCount());
        updateState();
    }


    public void releaseViewPager() {
        if (viewPager != null) {
            viewPager.removeOnPageChangeListener(this);
            viewPager = null;
        }
    }


    public void setRtlMode(@Nullable RtlMode mode) {
        Indicator indicator = manager.indicator();
        if (mode == null) {
            indicator.setRtlMode(RtlMode.Off);
        } else {
            indicator.setRtlMode(mode);
        }

        if (viewPager == null) {
            return;
        }

        int selectedPosition = indicator.getSelectedPosition();
        int position = selectedPosition;

        if (isRtl()) {
            position = (indicator.getCount() - 1) - selectedPosition;

        } else if (viewPager != null) {
            position = viewPager.getCurrentItem();
        }

        indicator.setLastSelectedPosition(position);
        indicator.setSelectingPosition(position);
        indicator.setSelectedPosition(position);
        invalidate();
    }


    public int getSelection() {
        return manager.indicator().getSelectedPosition();
    }


    public void setSelection(int position) {
        Indicator indicator = manager.indicator();
        position = adjustPosition(position);

        if (position == indicator.getSelectedPosition() || position == indicator.getSelectingPosition()) {
            return;
        }

        indicator.setInteractiveAnimation(false);
        indicator.setLastSelectedPosition(indicator.getSelectedPosition());
        indicator.setSelectingPosition(position);
        indicator.setSelectedPosition(position);
        manager.animate().basic();
    }

    public void setSelected(int position) {
        Indicator indicator = manager.indicator();
        AnimationType animationType = indicator.getAnimationType();
        indicator.setAnimationType(AnimationType.NONE);

        setSelection(position);
        indicator.setAnimationType(animationType);
    }


    public void clearSelection() {
        Indicator indicator = manager.indicator();
        indicator.setInteractiveAnimation(false);
        indicator.setLastSelectedPosition(Indicator.COUNT_NONE);
        indicator.setSelectingPosition(Indicator.COUNT_NONE);
        indicator.setSelectedPosition(Indicator.COUNT_NONE);
        manager.animate().basic();
    }


    public void setProgress(int selectingPosition, float progress) {
        Indicator indicator = manager.indicator();
        if (!indicator.isInteractiveAnimation()) {
            return;
        }

        int count = indicator.getCount();
        if (count <= 0 || selectingPosition < 0) {
            selectingPosition = 0;

        } else if (selectingPosition > count - 1) {
            selectingPosition = count - 1;
        }

        if (progress < 0) {
            progress = 0;

        } else if (progress > 1) {
            progress = 1;
        }

        if (progress == 1) {
            indicator.setLastSelectedPosition(indicator.getSelectedPosition());
            indicator.setSelectedPosition(selectingPosition);
        }

        indicator.setSelectingPosition(selectingPosition);
        manager.animate().interactive(progress);
    }

    public void setClickListener(@Nullable DrawController.ClickListener listener) {
        manager.drawer().setClickListener(listener);
    }

    private void init(@Nullable AttributeSet attrs) {
        setupId();
        initIndicatorManager(attrs);
    }

    private void setupId() {
        if (getId() == NO_ID) {
            setId(IdUtils.generateViewId());
        }
    }

    private void initIndicatorManager(@Nullable AttributeSet attrs) {
        manager = new IndicatorManager(this);
        manager.drawer().initAttributes(getContext(), attrs);

        Indicator indicator = manager.indicator();
        indicator.setPaddingLeft(getPaddingLeft());
        indicator.setPaddingTop(getPaddingTop());
        indicator.setPaddingRight(getPaddingRight());
        indicator.setPaddingBottom(getPaddingBottom());
        isInteractionEnabled = indicator.isInteractiveAnimation();
    }

    private void registerSetObserver() {
        if (setObserver != null || viewPager == null || viewPager.getAdapter() == null) {
            return;
        }

        setObserver = new DataSetObserver() {
            @Override
            public void onChanged() {
                updateState();
            }
        };

        try {
            viewPager.getAdapter().registerDataSetObserver(setObserver);
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    private void unRegisterSetObserver() {
        if (setObserver == null || viewPager == null || viewPager.getAdapter() == null) {
            return;
        }

        try {
            viewPager.getAdapter().unregisterDataSetObserver(setObserver);
            setObserver = null;
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    private void updateState() {
        if (viewPager == null || viewPager.getAdapter() == null) {
            return;
        }

        int count = viewPager.getAdapter().getCount();
        int selectedPos = isRtl() ? (count - 1) - viewPager.getCurrentItem() : viewPager.getCurrentItem();

        manager.indicator().setSelectedPosition(selectedPos);
        manager.indicator().setSelectingPosition(selectedPos);
        manager.indicator().setLastSelectedPosition(selectedPos);
        manager.indicator().setCount(count);
        manager.animate().end();

        updateVisibility();
        requestLayout();
    }

    private void updateVisibility() {
        if (!manager.indicator().isAutoVisibility()) {
            return;
        }

        int count = manager.indicator().getCount();
        int visibility = getVisibility();

        if (visibility != VISIBLE && count > Indicator.MIN_COUNT) {
            setVisibility(VISIBLE);

        } else if (visibility != INVISIBLE && count <= Indicator.MIN_COUNT) {
            setVisibility(View.INVISIBLE);
        }
    }

    private void onPageSelect(int position) {
        Indicator indicator = manager.indicator();
        boolean canSelectIndicator = isViewMeasured();
        int count = indicator.getCount();

        if (canSelectIndicator) {
            if (isRtl()) {
                position = (count - 1) - position;
            }

            setSelection(position);
        }
    }

	private void onPageScroll(int position, float positionOffset) {
        Indicator indicator = manager.indicator();
        AnimationType animationType = indicator.getAnimationType();
        boolean interactiveAnimation = indicator.isInteractiveAnimation();
        boolean canSelectIndicator = isViewMeasured() && interactiveAnimation && animationType != AnimationType.NONE;

        if (!canSelectIndicator) {
            return;
        }

        Pair<Integer, Float> progressPair = CoordinatesUtils.getProgress(indicator, position, positionOffset, isRtl());
        int selectingPosition = progressPair.first;
        float selectingProgress = progressPair.second;
        setProgress(selectingPosition, selectingProgress);
    }

    private boolean isRtl() {
        switch (manager.indicator().getRtlMode()) {
            case On:
                return true;

            case Off:
                return false;

            case Auto:
                return TextUtilsCompat.getLayoutDirectionFromLocale(getContext().getResources().getConfiguration().locale) == ViewCompat.LAYOUT_DIRECTION_RTL;
        }

        return false;
    }

    private boolean isViewMeasured() {
        return getMeasuredHeight() != 0 || getMeasuredWidth() != 0;
    }

    private void findViewPager(@Nullable ViewParent viewParent) {
        boolean isValidParent = viewParent != null &&
                viewParent instanceof ViewGroup &&
                ((ViewGroup) viewParent).getChildCount() > 0;

        if (!isValidParent) {
            return;
        }

        int viewPagerId = manager.indicator().getViewPagerId();
        ViewPager viewPager = findViewPager((ViewGroup) viewParent, viewPagerId);

        if (viewPager != null) {
            setViewPager(viewPager);
        } else {
            findViewPager(viewParent.getParent());
        }
    }

    @Nullable
    private ViewPager findViewPager(@NonNull ViewGroup viewGroup, int id) {
        if (viewGroup.getChildCount() <= 0) {
            return null;
        }

        View view = viewGroup.findViewById(id);
        if (view != null && view instanceof ViewPager) {
            return (ViewPager) view;
        } else {
            return null;
        }
    }

    private int adjustPosition(int position){
        Indicator indicator = manager.indicator();
        int count = indicator.getCount();
        int lastPosition = count - 1;

        if (position < 0) {
            position = 0;

        } else if (position > lastPosition) {
            position = lastPosition;
        }

        return position;
    }
}

File: autoimageslider/src/main/java/com/smarteist/autoimageslider/IndicatorView/PageIndicatorView.kt
package com.smarteist.autoimageslider.IndicatorView

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.Context
import android.database.DataSetObserver
import android.graphics.Canvas
import android.os.Build
import android.os.Parcelable
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.ViewParent
import androidx.core.text.TextUtilsCompat
import androidx.core.view.ViewCompat
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import com.smarteist.autoimageslider.IndicatorView.animation.type.IndicatorAnimationType
import com.smarteist.autoimageslider.IndicatorView.animation.type.ScaleAnimation
import com.smarteist.autoimageslider.IndicatorView.draw.controller.DrawController.ClickListener
import com.smarteist.autoimageslider.IndicatorView.draw.data.Indicator
import com.smarteist.autoimageslider.IndicatorView.draw.data.Orientation
import com.smarteist.autoimageslider.IndicatorView.draw.data.PositionSavedState
import com.smarteist.autoimageslider.IndicatorView.draw.data.RtlMode
import com.smarteist.autoimageslider.IndicatorView.utils.CoordinatesUtils.getProgress
import com.smarteist.autoimageslider.IndicatorView.utils.DensityUtils.dpToPx
import com.smarteist.autoimageslider.IndicatorView.utils.IdUtils
import com.smarteist.autoimageslider.InfiniteAdapter.InfinitePagerAdapter
import com.smarteist.autoimageslider.SliderPager

class PageIndicatorView : View, SliderPager.OnPageChangeListener, IndicatorManager.Listener, SliderPager.OnAdapterChangeListener {
    private var manager: IndicatorManager? = null
    private var setObserver: DataSetObserver? = null
    private var viewPager: SliderPager? = null
    private var isInteractionEnabled = false

    constructor(context: Context?) : super(context) {
        init(null)
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init(attrs)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(attrs)
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(attrs)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        findViewPager(parent)
    }

    override fun onDetachedFromWindow() {
        unRegisterSetObserver()
        super.onDetachedFromWindow()
    }

    public override fun onSaveInstanceState(): Parcelable? {
        val indicator = manager!!.indicator()
        val positionSavedState = PositionSavedState(super.onSaveInstanceState())
        positionSavedState.selectedPosition = indicator.selectedPosition
        positionSavedState.selectingPosition = indicator.selectingPosition
        positionSavedState.lastSelectedPosition = indicator.lastSelectedPosition
        return positionSavedState
    }

    public override fun onRestoreInstanceState(state: Parcelable) {
        if (state is PositionSavedState) {
            val indicator = manager!!.indicator()
            val positionSavedState = state
            indicator.selectedPosition = positionSavedState.selectedPosition
            indicator.selectingPosition = positionSavedState.selectingPosition
            indicator.lastSelectedPosition = positionSavedState.lastSelectedPosition
            super.onRestoreInstanceState(positionSavedState.superState)
        } else {
            super.onRestoreInstanceState(state)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val pair = manager!!.drawer().measureViewSize(widthMeasureSpec, heightMeasureSpec)
        setMeasuredDimension(pair.first, pair.second)
    }

    override fun onDraw(canvas: Canvas) {
        manager!!.drawer().draw(canvas)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        manager!!.drawer().touch(event)
        return true
    }

    override fun onIndicatorUpdated() {
        invalidate()
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        onPageScroll(position, positionOffset)
    }

    override fun onPageSelected(position: Int) {
        onPageSelect(position)
    }

    override fun onPageScrollStateChanged(state: Int) {
        if (state == ViewPager.SCROLL_STATE_IDLE) {
            manager!!.indicator().isInteractiveAnimation = isInteractionEnabled
        }
    }

    override fun onAdapterChanged(viewPager: SliderPager, oldAdapter: PagerAdapter?, newAdapter: PagerAdapter?) {
        updateState()
    }

    var count: Int
        get() = manager!!.indicator().count
        set(count) {
            if (count >= 0 && manager!!.indicator().count != count) {
                manager!!.indicator().count = count
                updateVisibility()
                requestLayout()
            }
        }

    fun setDynamicCount(dynamicCount: Boolean) {
        manager!!.indicator().isDynamicCount = dynamicCount
        if (dynamicCount) {
            registerSetObserver()
        } else {
            unRegisterSetObserver()
        }
    }

    fun setRadius(radiusPx: Float) {
        var radiusPx = radiusPx
        if (radiusPx < 0) {
            radiusPx = 0f
        }
        manager!!.indicator().radius = radiusPx.toInt()
        invalidate()
    }

    var radius: Int
        get() = manager!!.indicator().radius
        set(radiusDp) {
            var radiusDp = radiusDp
            if (radiusDp < 0) {
                radiusDp = 0
            }
            val radiusPx = dpToPx(radiusDp)
            manager!!.indicator().radius = radiusPx
            invalidate()
        }

    fun setPadding(paddingPx: Float) {
        var paddingPx = paddingPx
        if (paddingPx < 0) {
            paddingPx = 0f
        }
        manager!!.indicator().padding = paddingPx.toInt()
        invalidate()
    }

    var padding: Int
        get() = manager!!.indicator().padding
        set(paddingDp) {
            var paddingDp = paddingDp
            if (paddingDp < 0) {
                paddingDp = 0
            }
            val paddingPx = dpToPx(paddingDp)
            manager!!.indicator().padding = paddingPx
            invalidate()
        }
    var scaleFactor: Float
        get() = manager!!.indicator().scaleFactor
        set(factor) {
            var factor = factor
            if (factor > ScaleAnimation.MAX_SCALE_FACTOR) {
                factor = ScaleAnimation.MAX_SCALE_FACTOR
            } else if (factor < ScaleAnimation.MIN_SCALE_FACTOR) {
                factor = ScaleAnimation.MIN_SCALE_FACTOR
            }
            manager!!.indicator().scaleFactor = factor
        }

    fun setStrokeWidth(strokePx: Float) {
        var strokePx = strokePx
        val radiusPx = manager!!.indicator().radius
        if (strokePx < 0) {
            strokePx = 0f
        } else if (strokePx > radiusPx) {
            strokePx = radiusPx.toFloat()
        }
        manager!!.indicator().strokeHere = strokePx.toInt()
        invalidate()
    }

    fun setStrokeWidth(strokeDp: Int) {
        var strokePx = dpToPx(strokeDp)
        val radiusPx = manager!!.indicator().radius
        if (strokePx < 0) {
            strokePx = 0
        } else if (strokePx > radiusPx) {
            strokePx = radiusPx
        }
        manager!!.indicator().strokeHere = (strokePx)
        invalidate()
    }

    val strokeWidth: Int get() = manager!!.indicator().strokeHere
    var selectedColor: Int
        get() = manager!!.indicator().selectedColor
        set(color) {
            manager!!.indicator().selectedColor = color
            invalidate()
        }
    var unselectedColor: Int
        get() = manager!!.indicator().unselectedColor
        set(color) {
            manager!!.indicator().unselectedColor = color
            invalidate()
        }

    fun setAutoVisibility(autoVisibility: Boolean) {
        if (!autoVisibility) {
            visibility = VISIBLE
        }
        manager!!.indicator().isAutoVisibility = autoVisibility
        updateVisibility()
    }

    fun setOrientation(orientation: Orientation?) {
        if (orientation != null) {
            manager!!.indicator().orientation = orientation
            requestLayout()
        }
    }

    var animationDuration: Long
        get() = manager!!.indicator().animationDuration
        set(duration) {
            manager!!.indicator().animationDuration = duration
        }

    fun setAnimationType(type: IndicatorAnimationType?) {
        manager!!.onValueUpdated(null)
        if (type != null) {
            manager!!.indicator().animationType = type
        } else {
            manager!!.indicator().animationType = IndicatorAnimationType.NONE
        }
        invalidate()
    }

    fun setInteractiveAnimation(isInteractive: Boolean) {
        manager!!.indicator().isInteractiveAnimation = isInteractive
        isInteractionEnabled = isInteractive
    }

    fun setViewPager(pager: SliderPager?) {
        releaseViewPager()
        if (pager == null) {
            return
        }
        viewPager = pager
        viewPager!!.addOnPageChangeListener(this)
        viewPager!!.addOnAdapterChangeListener(this)
        manager!!.indicator().viewPagerId = viewPager!!.id
        setDynamicCount(manager!!.indicator().isDynamicCount)
        updateState()
    }

    fun releaseViewPager() {
        if (viewPager != null) {
            viewPager!!.removeOnPageChangeListener(this)
            viewPager = null
        }
    }

    fun setRtlMode(mode: RtlMode?) {
        val indicator = manager!!.indicator()
        if (mode == null) {
            indicator.rtlMode = RtlMode.Off
        } else {
            indicator.rtlMode = mode
        }
        if (viewPager == null) {
            return
        }
        val selectedPosition = indicator.selectedPosition
        var position = selectedPosition
        if (isRtl) {
            position = indicator.count - 1 - selectedPosition
        } else if (viewPager != null) {
            position = viewPager!!.currentItem
        }
        indicator.lastSelectedPosition = position
        indicator.selectingPosition = position
        indicator.selectedPosition = position
        invalidate()
    }

    var selection: Int
        get() = manager!!.indicator().selectedPosition
        set(position) {
            var position = position
            val indicator = manager!!.indicator()
            position = adjustPosition(position)
            if (position == indicator.selectedPosition || position == indicator.selectingPosition) {
                return
            }
            indicator.isInteractiveAnimation = false
            indicator.lastSelectedPosition = indicator.selectedPosition
            indicator.selectingPosition = position
            indicator.selectedPosition = position
            manager!!.animate().basic()
        }

    fun setSelected(position: Int) {
        val indicator = manager!!.indicator()
        val animationType = indicator.animationType
        indicator.animationType = IndicatorAnimationType.NONE
        selection = position
        indicator.animationType = animationType
    }

    fun clearSelection() {
        val indicator = manager!!.indicator()
        indicator.isInteractiveAnimation = false
        indicator.lastSelectedPosition = Indicator.COUNT_NONE
        indicator.selectingPosition = Indicator.COUNT_NONE
        indicator.selectedPosition = Indicator.COUNT_NONE
        manager!!.animate().basic()
    }

    fun setProgress(selectingPosition: Int, progress: Float) {
        var selectingPosition = selectingPosition
        var progress = progress
        val indicator = manager!!.indicator()
        if (!indicator.isInteractiveAnimation) {
            return
        }
        val count = indicator.count
        if (count <= 0 || selectingPosition < 0) {
            selectingPosition = 0
        } else if (selectingPosition > count - 1) {
            selectingPosition = count - 1
        }
        if (progress < 0) {
            progress = 0f
        } else if (progress > 1) {
            progress = 1f
        }
        if (progress == 1f) {
            indicator.lastSelectedPosition = indicator.selectedPosition
            indicator.selectedPosition = selectingPosition
        }
        indicator.selectingPosition = selectingPosition
        manager!!.animate().interactive(progress)
    }

    fun setClickListener(listener: ClickListener?) {
        manager!!.drawer().setClickListener(listener)
    }

    private fun init(attrs: AttributeSet?) {
        setupId()
        initIndicatorManager(attrs)
    }

    private fun setupId() {
        if (id == NO_ID) {
            id = IdUtils.generateViewId()
        }
    }

    private fun initIndicatorManager(attrs: AttributeSet?) {
        manager = IndicatorManager(this)
        manager!!.drawer().initAttributes(context, attrs)
        val indicator = manager!!.indicator()
        indicator.paddingLeft = paddingLeft
        indicator.paddingTop = paddingTop
        indicator.paddingRight = paddingRight
        indicator.paddingBottom = paddingBottom
        isInteractionEnabled = indicator.isInteractiveAnimation
    }

    private fun registerSetObserver() {
        if (setObserver != null || viewPager == null || viewPager!!.adapter == null) {
            return
        }
        setObserver = object : DataSetObserver() {
            override fun onChanged() {
                updateState()
            }
        }
        try {
            viewPager!!.adapter!!.registerDataSetObserver(setObserver!!)
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }
    }

    private fun unRegisterSetObserver() {
        if (setObserver == null || viewPager == null || viewPager!!.adapter == null) {
            return
        }
        try {
            viewPager!!.adapter!!.unregisterDataSetObserver(setObserver!!)
            setObserver = null
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }
    }

    private fun updateState() {
        if (viewPager == null || viewPager!!.adapter == null) {
            return
        }
        val count: Int
        val position: Int
        if (viewPager!!.adapter is InfinitePagerAdapter) {
            count = (viewPager!!.adapter as InfinitePagerAdapter?)!!.realCount
            position = if (count > 0) {
                viewPager!!.currentItem % count
            } else {
                0
            }
        } else {
            count = viewPager!!.adapter!!.count
            position = viewPager!!.currentItem
        }
        val selectedPos = if (isRtl) count - 1 - position else position
        manager!!.indicator().selectedPosition = selectedPos
        manager!!.indicator().selectingPosition = selectedPos
        manager!!.indicator().lastSelectedPosition = selectedPos
        manager!!.indicator().count = count
        manager!!.animate().end()
        updateVisibility()
        requestLayout()
    }

    private fun updateVisibility() {
        if (!manager!!.indicator().isAutoVisibility) {
            return
        }
        val count = manager!!.indicator().count
        val visibility = visibility
        if (visibility != VISIBLE && count > Indicator.MIN_COUNT) {
            setVisibility(VISIBLE)
        } else if (visibility != INVISIBLE && count <= Indicator.MIN_COUNT) {
            setVisibility(INVISIBLE)
        }
    }

    private fun onPageSelect(position: Int) {
        var position = position
        val indicator = manager!!.indicator()
        val canSelectIndicator = isViewMeasured
        val count = indicator.count
        if (canSelectIndicator) {
            if (isRtl) {
                position = count - 1 - position
            }
            selection = position
        }
    }

    private fun onPageScroll(position: Int, positionOffset: Float) {
        val indicator = manager!!.indicator()
        val animationType = indicator.animationType
        val interactiveAnimation = indicator.isInteractiveAnimation
        val canSelectIndicator = isViewMeasured && interactiveAnimation && animationType !== IndicatorAnimationType.NONE
        if (!canSelectIndicator) {
            return
        }
        val progressPair = getProgress(indicator, position, positionOffset, isRtl)
        val selectingPosition = progressPair.first
        val selectingProgress = progressPair.second
        setProgress(selectingPosition, selectingProgress)
    }

    private val isRtl: Boolean
        private get() {
            when (manager!!.indicator().rtlMode) {
                RtlMode.On -> return true
                RtlMode.Off -> return false
                RtlMode.Auto -> return TextUtilsCompat.getLayoutDirectionFromLocale(context.resources.configuration.locale) == ViewCompat.LAYOUT_DIRECTION_RTL
            }
            return false
        }
    private val isViewMeasured: Boolean
        private get() = measuredHeight != 0 || measuredWidth != 0

    private fun findViewPager(viewParent: ViewParent?) {
        val isValidParent = viewParent != null &&
                viewParent is ViewGroup && viewParent.childCount > 0
        if (!isValidParent) {
            return
        }
        val viewPagerId = manager!!.indicator().viewPagerId
        val viewPager = findViewPager((viewParent as ViewGroup?)!!, viewPagerId)
        if (viewPager != null) {
            setViewPager(viewPager)
        } else {
            findViewPager(viewParent!!.parent)
        }
    }

    private fun findViewPager(viewGroup: ViewGroup, id: Int): SliderPager? {
        if (viewGroup.childCount <= 0) {
            return null
        }
        val view = viewGroup.findViewById<View>(id)
        return if (view != null && view is SliderPager) {
            view
        } else {
            null
        }
    }

    private fun adjustPosition(position: Int): Int {
        var position = position
        val indicator = manager!!.indicator()
        val count = indicator.count
        val lastPosition = count - 1
        if (position <= 0) {
            position = 0
        } else if (position > lastPosition) {
            position = lastPosition
        }
        return position
    }
}